apm-collector/apm-collector-storage/collector-storage-shardingjdbc-provider/src/main/java/org/apache/skywalking/apm/collector/storage/shardingjdbc/base/dao/AbstractPersistenceShardingjdbcDAO.java

```diff
@@ -0,0 +1,119 @@
+ */
+public abstract class AbstractPersistenceShardingjdbcDAO<STREAM_DATA extends StreamData> extends ShardingjdbcDAO implements IPersistenceDAO<ShardingjdbcSqlEntity, ShardingjdbcSqlEntity, STREAM_DATA> {
+
+    private final Logger logger = LoggerFactory.getLogger(AbstractPersistenceShardingjdbcDAO.class);
+
+    public AbstractPersistenceShardingjdbcDAO(ShardingjdbcClient client) {
+        super(client);
+    }
+
+    private static final String GET_SQL = "select * from {0} where {1} = ?";
+
+    protected abstract STREAM_DATA shardingjdbcDataToStreamData(ResultSet resultSet) throws SQLException;
+
+    protected abstract String tableName();
+
+    @Override public STREAM_DATA get(String id) {
+        String sql = SqlBuilder.buildSql(GET_SQL, tableName(), CommonTable.ID.getName());
+
+        Object[] params = new Object[] {id};
+        try (
+                ResultSet rs = getClient().executeQuery(sql, params);
+                Statement statement = rs.getStatement();
+                Connection conn = statement.getConnection();
+            ) {
+            if (rs.next()) {
+                return shardingjdbcDataToStreamData(rs);
+            }
+        } catch (SQLException | ShardingjdbcClientException e) {
+            logger.error(e.getMessage(), e);
+        }
+        return null;
+    }
+
+    protected abstract Map<String, Object> streamDataToShardingjdbcData(STREAM_DATA streamData);
+
+    @Override public final ShardingjdbcSqlEntity prepareBatchInsert(STREAM_DATA streamData) {
+        Map<String, Object> source = streamDataToShardingjdbcData(streamData);
+        source.put(CommonTable.ID.getName(), streamData.getId());
+
+        ShardingjdbcSqlEntity entity = new ShardingjdbcSqlEntity();
+
+        String sql = SqlBuilder.buildBatchInsertSql(tableName(), source.keySet());
+        entity.setSql(sql);
+        entity.setParams(source.values().toArray(new Object[0]));
+        return entity;
+    }
+
+    @Override public final ShardingjdbcSqlEntity prepareBatchUpdate(STREAM_DATA streamData) {
+        Map<String, Object> source = streamDataToShardingjdbcData(streamData);
+
+        ShardingjdbcSqlEntity entity = new ShardingjdbcSqlEntity();
+        String sql = SqlBuilder.buildBatchUpdateSql(tableName(), source.keySet(), CommonTable.ID.getName());
+        entity.setSql(sql);
+        List<Object> values = new ArrayList<>(source.values());
+        values.add(streamData.getId());
+        entity.setParams(values.toArray(new Object[0]));
+        return entity;
+    }
+    
+    protected abstract String timeBucketColumnNameForDelete();
+
+    @Override public void deleteHistory(Long timeBucketBefore) {
+        ShardingjdbcClient client = getClient();
+        
+        String dynamicSql = "delete from {0} where {1} <= ?";
+        String sql = SqlBuilder.buildSql(dynamicSql, tableName(), timeBucketColumnNameForDelete());
+        
+        long timeBucket = TimeBucketUtils.INSTANCE.getMinuteTimeBucket(timeBucketBefore);
```

Comment:
Why transform timeBucketBefore to  minute time bucket?

Reply:
I have modified it and other classes.